git/gogit/repository_client.go

```diff
@@ -0,0 +1,318 @@
+/*
+Copyright 2022 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gogit
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/ProtonMail/go-crypto/openpgp"
+	"github.com/go-git/go-billy/v5"
+	"github.com/go-git/go-billy/v5/memfs"
+	"github.com/go-git/go-billy/v5/osfs"
+	extgogit "github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/config"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/cache"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/go-git/go-git/v5/storage"
+	"github.com/go-git/go-git/v5/storage/filesystem"
+	"github.com/go-git/go-git/v5/storage/memory"
+
+	"github.com/fluxcd/pkg/git"
+)
+
+// GoGitClient implements git.GitClient
+type GoGitClient struct {
+	path       string
+	repository *extgogit.Repository
+	authOpts   *git.AuthOptions
+	storer     storage.Storer
+	worktreeFS billy.Filesystem
+}
+
+var _ git.RepositoryClient = &GoGitClient{}
+
+type ClientOption func(*GoGitClient) error
+
+// NewGoGitClient returns a new GoGitClient.
+func NewGoGitClient(path string, authOpts *git.AuthOptions, clientOpts ...ClientOption) (*GoGitClient, error) {
+	g := &GoGitClient{
+		path:     path,
+		authOpts: authOpts,
+	}
+
+	if len(clientOpts) == 0 {
+		clientOpts = append(clientOpts, UseDiskStorage)
+	}
+
+	for _, clientOpt := range clientOpts {
+		if err := clientOpt(g); err != nil {
+			return nil, err
+		}
+	}
+
+	return g, nil
+}
+
```

Comment:
```suggestion
func WithStorer(s storage.Storer) ClientOption {
	return func(c *GoGitClient) error {
		c.storer = s
	}
}

func WithWorkTreeFS(wt billy.Filesystem) ClientOption {
	return func(c *GoGitClient) error {
		c.worktreeFS = wt
	}
}
```

---

Or the alternative:

```go
type ClientOption interface {
	apply(c *GoGitClient) error
}

type WithStorer storage.Storer

func(o WithStorer) apply(c *GoGitClient) error {
	c.storer = storage.Storer(o)
}
```

Reply:
I would +1 on the `WithStorer` and `WithWorkTreeFS` approach.

Reply:
I would still keep the other methods for consumer convenience, and in addition, ensure that at least both are set at the end of the init. As it's now possible to pass an option that does not set both.